syntax = "proto3";

package lambda_action;

// Lambda Action Service
service LambdaAction {
  rpc InvokeFunction(InvokeFunctionRequest) returns (InvokeFunctionResponse);
  rpc InvokeAsync(InvokeAsyncRequest) returns (InvokeAsyncResponse);
  rpc BatchInvoke(BatchInvokeRequest) returns (BatchInvokeResponse);
  rpc ListFunctions(ListFunctionsRequest) returns (ListFunctionsResponse);
  rpc GetFunctionConfig(GetFunctionConfigRequest) returns (GetFunctionConfigResponse);
}

// Invocation structures
message InvokeFunctionRequest {
  string function_name = 1;
  string payload = 2;
  string invocation_type = 3;  // RequestResponse, Event, DryRun
  string alias = 4;
  string version = 5;
  string token = 6;  // JWT token
}

message InvokeFunctionResponse {
  bool success = 1;
  int32 status_code = 2;
  string payload = 3;
  string function_error = 4;
  string log_result = 5;
  string executed_version = 6;
  string error = 7;
}

message InvokeAsyncRequest {
  string function_name = 1;
  string payload = 2;
  string token = 3;
}

message InvokeAsyncResponse {
  bool success = 1;
  int32 status_code = 2;
  string request_id = 3;
  string error = 4;
}

message BatchInvokeRequest {
  repeated InvokeFunctionRequest invocations = 1;
  string token = 2;
}

message BatchInvokeResponse {
  repeated InvokeFunctionResponse responses = 1;
}

message ListFunctionsRequest {
  string token = 1;
  int32 max_items = 2;
  string next_marker = 3;
}

message ListFunctionsResponse {
  bool success = 1;
  repeated FunctionInfo functions = 2;
  string next_marker = 3;
  string error = 4;
}

message FunctionInfo {
  string function_name = 1;
  string function_arn = 2;
  string runtime = 3;
  string role = 4;
  string handler = 5;
  int32 code_size = 6;
  string description = 7;
  int32 timeout = 8;
  int32 memory_size = 9;
  string last_modified = 10;
  string version = 11;
}

message GetFunctionConfigRequest {
  string function_name = 1;
  string token = 2;
}

message GetFunctionConfigResponse {
  bool success = 1;
  FunctionInfo function_info = 2;
  string error = 3;
}
